src/coreclr/pal/inc/unixasmmacrosriscv64.inc

```diff
@@ -1,42 +1,328 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#error "TODO-RISCV64: review this; missing many macros for VM"
-
 .macro NESTED_ENTRY Name, Section, Handler
-        LEAF_ENTRY \Name, \Section
-        .ifnc \Handler, NoHandler
-        .personality C_FUNC(\Handler)
-        .endif
+    LEAF_ENTRY \Name, \Section
+    .ifnc \Handler, NoHandler
+        .cfi_personality 0x1c, C_FUNC(\Handler) // 0x1c == DW_EH_PE_pcrel | DW_EH_PE_sdata8
+    .endif
 .endm
 
 .macro NESTED_END Name, Section
-        LEAF_END \Name, \Section
+    LEAF_END \Name, \Section
 .endm
 
 .macro PATCH_LABEL Name
-        .global C_FUNC(\Name)
+    .global C_FUNC(\Name)
 C_FUNC(\Name):
 .endm
 
 .macro LEAF_ENTRY Name, Section
-        .global C_FUNC(\Name)
-        .type \Name, %function
+    .global C_FUNC(\Name)
+    .type \Name, %function
 C_FUNC(\Name):
-        .cfi_startproc
+    .cfi_startproc
 .endm
 
 .macro LEAF_END Name, Section
-        .size \Name, .-\Name
-        .cfi_endproc
+    .size \Name, .-\Name
+    .cfi_endproc
 .endm
 
 .macro LEAF_END_MARKED Name, Section
 C_FUNC(\Name\()_End):
-        .global C_FUNC(\Name\()_End)
-        LEAF_END \Name, \Section
+    .global C_FUNC(\Name\()_End)
+    LEAF_END \Name, \Section
+    // make sure this symbol gets its own address
+    nop
+.endm
+
+.macro PREPARE_EXTERNAL_VAR Name, HelperReg
+    lla  \HelperReg, \Name
+.endm
+
+.macro PROLOG_STACK_ALLOC Size
+    addi sp, sp, -\Size
+    //.cfi_adjust_cfa_offset \Size
+    .cfi_def_cfa sp,\Size
+.endm
+
+.macro EPILOG_STACK_FREE Size
+    addi sp, sp, \Size
+    //.cfi_adjust_cfa_offset -\Size
+    .cfi_def_cfa sp,-\Size
```

Comment:
Is there any reason for not using the `.cfi_adjust_cfa_offset`? I am just curious, it is not a problem.

Reply:
I implemented based on LOONGARCH. I updated assembly codes and copied most other implementations (including cfi).